feat(hadoop): Add hadoop products data source

jungchanSon opened the pull request:
## Description
- add `hadoop_products.md`
- add `hadoop_products_data_source.go`
- add `hadoop_products_data_source_test.go`
- add `hadoop_products_data_source` in provider

## Test
```
$ go test ./internal/service/hadoop/hadoop_products_data_source_test.go -v
=== RUN   TestAccDataSourceNcloudHadoopProducts_basic
--- PASS: TestAccDataSourceNcloudHadoopProducts_basic (3.05s)
PASS
ok      command-line-arguments  3.883s
```

jungchanSon committed:
feat(hadoop): add hadoop data source

jungchanSon committed:
docs(hadoop): add hadoop_products.md

jungchanSon committed:
refactor(hadoop): change function and type code order

jungchanSon committed:
Merge branch 'main' into feat/hadoop_products

youngmn commented on docs/data-sources/hadoop_products.md:
```diff
@@ -0,0 +1,74 @@
+The following arguments are supported:
+
+* `image_product_code` - (Required) You can get one from `data.ncloud_hadoop_images`. This is a required value, and each available Hadoop's specification varies depending on the hadoop image product.
+* `product_code` - (Required) Product code you want to view on the list. Use this when searching for 1 product.
```
사용자가 대부분 product_code 값을 알기 위해 사용할 것인데, product_code 를 입력 argument 로 제공할 필요가 없어 보입니다. filter 로도 특정 product_code 를 조회할 수 있으므로 불필요해 보입니다.

youngmn commented on docs/data-sources/hadoop_products.md:
```diff
@@ -0,0 +1,74 @@
+* `image_product_code` - (Required) You can get one from `data.ncloud_hadoop_images`. This is a required value, and each available Hadoop's specification varies depending on the hadoop image product.
+* `product_code` - (Required) Product code you want to view on the list. Use this when searching for 1 product.
+* `infra_resource_detail_type_code` - (Optional) Hadoop Other Server infra Detailed Product Code. Options : MSTDT | EDGND
+* `exclusion_product_code` - (Optional) Product code you want to exclude on the list.
```
product_code 와 같은 이유로 불필요 합니다.

youngmn commented on internal/service/hadoop/hadoop_products_data_source.go:
```diff
@@ -0,0 +1,276 @@
+			"image_product_code": schema.StringAttribute{
+				Required: true,
+			},
+			"product_code": schema.StringAttribute{
```
product_code 는 삭제

youngmn commented on internal/service/hadoop/hadoop_products_data_source.go:
```diff
@@ -0,0 +1,276 @@
+			"infra_resource_detail_type_code": schema.StringAttribute{
+				Optional: true,
+			},
+			"exclusion_product_code": schema.StringAttribute{
```
exclusion_product_code 삭제

youngmn commented on internal/service/hadoop/hadoop_products_data_source.go:
```diff
@@ -0,0 +1,276 @@
+}
+
+func (h *hadoopProductsDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data hadoopProductsDataSourceModel
```
VPC 에서만 사용 가능하므로 예외처리를 추가하면 좋을 듯 합니다

youngmn commented on internal/service/hadoop/hadoop_products_data_source.go:
```diff
@@ -0,0 +1,276 @@
+
+	if !data.ImageProductCode.IsNull() && !data.ImageProductCode.IsUnknown() {
+		reqParams.CloudHadoopImageProductCode = data.ImageProductCode.ValueStringPointer()
+	}
```
image_product_code 는 Requied 이므로 Null, Unknown 에 해당되지 않습니다. RegionCode 와 같이 한줄로 값을 설정해도 됩니다

youngmn commented on internal/service/hadoop/hadoop_products_data_source.go:
```diff
@@ -0,0 +1,276 @@
+
+	if !data.ProductCode.IsNull() && !data.ProductCode.IsUnknown() {
+		reqParams.ProductCode = data.ProductCode.ValueStringPointer()
+	}
```
product_code 는 삭제

youngmn commented on internal/service/hadoop/hadoop_products_data_source.go:
```diff
@@ -0,0 +1,276 @@
+
+	if !data.ExclusionProductCode.IsNull() && !data.ExclusionProductCode.IsUnknown() {
+		reqParams.ExclusionProductCode = data.ExclusionProductCode.ValueStringPointer()
+	}
```
exclusion_product_code 삭제

youngmn commented on internal/service/hadoop/hadoop_products_data_source.go:
```diff
@@ -0,0 +1,276 @@
+
+	tflog.Info(ctx, "GetHadoopProductList", map[string]any{
+		"reqParams": common.MarshalUncheckedString(reqParams),
+	})
```
이와 같이 로그를 출력하면 GetHadoopProductList tf_req_id=f821953a-9535-da14-d19d-6b200b46f661 @module=ncloud reqPrarams 와 같이 중간에 다른 로그가 추가로 붙게되어 알아보기 힘듭니다.
아래와 같이 이어서 출력하도록 수정하면 좋을 듯 합니다.
tflog.Info(ctx, "GetHadoopProductsList reqParams="+common.MarshalUncheckedString(reqParams))

youngmn commented on internal/service/hadoop/hadoop_products_data_source.go:
```diff
@@ -0,0 +1,276 @@
+			"GetHadoopProductList",
+			fmt.Sprintf("error: %s, reqParams: %s", err.Error(), common.MarshalUncheckedString(reqParams)),
+		)
+		resp.Diagnostics.Append(diags...)
```
CDB 상품군 전체의 에러 포맷을 개선한 것과 같이, 아래와 같이 한줄로 통일하면 좋을 것 같습니다.
resp.Diagnostics.AddError("READING ERROR", err.Error())

youngmn commented on internal/service/hadoop/hadoop_products_data_source.go:
```diff
@@ -0,0 +1,276 @@
+	}
+	tflog.Info(ctx, "GetHadoopProductList response", map[string]any{
+		"productResponse": common.MarshalUncheckedString(productsResp),
+	})
```
로그 포맷 변경 필요

youngmn commented on internal/service/hadoop/hadoop_products_data_source.go:
```diff
@@ -0,0 +1,276 @@
+	tflog.Info(ctx, "GetHadoopProductList response", map[string]any{
+		"productResponse": common.MarshalUncheckedString(productsResp),
+	})
+
```
productsResp.ProductList 의 조회 결과가 nil 이거나 list 가 0 인 경우에 대한 예외처리를 추가하여 panic 을 방지하면 좋을 듯 합니다.

youngmn commented on internal/service/hadoop/hadoop_products_data_source.go:
```diff
@@ -0,0 +1,276 @@
+			return
+		}
+	}
+	data.ID = types.StringValue(time.Now().UTC().String())
```
여기서 id 는 사용되지 않으므로 "" 빈값으로 설정합니다

youngmn commented on internal/service/hadoop/hadoop_products_data_source.go:
```diff
@@ -0,0 +1,276 @@
+		if err := json.Unmarshal([]byte(product.String()), &hadoopProduct); err != nil {
+			diags.AddError(
+				"Unmarshal",
+				fmt.Sprintf("error: %s", err.Error()),
```
에러 포맷을 통일하기 위해 아래와 같이 변경하면 좋을 듯 합니다.
"OUTPUT FILE ERROR", err.Error()

youngmn commented on internal/service/hadoop/hadoop_products_data_source.go:
```diff
@@ -0,0 +1,276 @@
+	if err := common.WriteToFile(path, hadoopProducts); err != nil {
+		diags.AddError(
+			"WriteToFile",
+			fmt.Sprintf("error: %s", err.Error()),
```
에러 포맷을 통일하기 위해 아래와 같이 변경하면 좋을 듯 합니다.
"OUTPUT FILE ERROR", err.Error()

youngmn commented on internal/service/hadoop/hadoop_products_data_source.go:
```diff
@@ -0,0 +1,276 @@
+type hadoopProductsDataSourceModel struct {
+	ID                          types.String `tfsdk:"id"`
+	ImageProductCode            types.String `tfsdk:"image_product_code"`
+	ProductCode                 types.String `tfsdk:"product_code"`
```
삭제

youngmn commented on internal/service/hadoop/hadoop_products_data_source.go:
```diff
@@ -0,0 +1,276 @@
+	ImageProductCode            types.String `tfsdk:"image_product_code"`
+	ProductCode                 types.String `tfsdk:"product_code"`
+	InfraResourceDetailTypeCode types.String `tfsdk:"infra_resource_detail_type_code"`
+	ExclusionProductCode        types.String `tfsdk:"exclusion_product_code"`
```
삭제

youngmn commented on docs/data-sources/hadoop_products.md:
````diff
@@ -0,0 +1,74 @@
+
+```terraform
+data "ncloud_hadoop_products" "all" {
+  image_product_code = "SW.VCHDP.LNX64.CNTOS.0708.HDP.15.B050"
````
상품코드는 deprecated 될 수 있는 old version 대신 new 버전으로 예제를 작성해 주면 좋을 듯 합니다. 
SW.VCHDP.LNX64.CNTOS.0708.HDP.21.B050

youngmn commented on internal/service/hadoop/hadoop_products_data_source_test.go:
```diff
@@ -0,0 +1,41 @@
+)
+
+func TestAccDataSourceNcloudHadoopProducts_basic(t *testing.T) {
+	imageProductCode := "SW.VCHDP.LNX64.CNTOS.0708.HDP.15.B050"
```
최신버전으로 변경
SW.VCHDP.LNX64.CNTOS.0708.HDP.21.B050

youngmn commented on internal/service/hadoop/hadoop_products_data_source_test.go:
```diff
@@ -0,0 +1,41 @@
+
+func TestAccDataSourceNcloudHadoopProducts_basic(t *testing.T) {
+	imageProductCode := "SW.VCHDP.LNX64.CNTOS.0708.HDP.15.B050"
+	productCode := "SVR.VCHDP.EDGND.HIMEM.C004.M032.NET.HDD.B050.G002"
```
삭제

youngmn commented on internal/service/hadoop/hadoop_products_data_source_test.go:
```diff
@@ -0,0 +1,41 @@
+		ProtoV6ProviderFactories: acctest.ProtoV6ProviderFactories,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataSourceHadoopProductsConfig(imageProductCode, productCode, infraResourceDetailTypeCode),
```
productCode 삭제

youngmn commented on internal/service/hadoop/hadoop_products_data_source_test.go:
```diff
@@ -0,0 +1,41 @@
+	return fmt.Sprintf(`
+data "ncloud_hadoop_products" "products" {
+	image_product_code = "%[1]s"
+	product_code = "%[2]s"
```
삭제

youngmn commented on internal/service/hadoop/hadoop_products_data_source_test.go:
```diff
@@ -0,0 +1,41 @@
+	})
+}
+
+func testAccDataSourceHadoopProductsConfig(imageProductCode, productCode, infraResourceDetailTypeCode string) string {
```
productCode 삭제

youngmn commented on internal/service/hadoop/hadoop_products_data_source_test.go:
```diff
@@ -0,0 +1,41 @@
+	product_code = "%[2]s"
+	infra_resource_detail_type_code = "%[3]s"
+}
+`, imageProductCode, productCode, infraResourceDetailTypeCode)
```
productCode 삭제

youngmn commented on internal/service/hadoop/hadoop_products_data_source_test.go:
```diff
@@ -0,0 +1,41 @@
+				Config: testAccDataSourceHadoopProductsConfig(imageProductCode, productCode, infraResourceDetailTypeCode),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("data.ncloud_hadoop_products.products", "product_list.0.infra_resource_type", "VCHDP"),
+					resource.TestCheckResourceAttr("data.ncloud_hadoop_products.products", "product_list.0.product_code", productCode),
```
productCode 삭제

youngmn reviewed (commented):
framework 로 개발된 CDB 상품군에 리팩토링한 내역들이 대부분 반영되어 있지 않아 관련 내용 위주로 리뷰 했습니다. 확인 부탁 드려요~

jungchanSon committed:
docs(hadoop): update 'image_product_code' and drop argument

jungchanSon committed:
refactor(hadoop): Update hadoop products data source

drop `product_code` and `exclude_product_code`
change errors and logs format

jungchanSon committed:
test(hadoop): Update hadoop_products_data_source_test.go

 change image product code
 drop product code and add product type
 add `filter` in test case

jungchanSon commented:
> framework 로 개발된 CDB 상품군에 리팩토링한 내역들이 대부분 반영되어 있지 않아 관련 내용 위주로 리뷰 했습니다. 확인 부탁 드려요~

리뷰 감사합니다 
말씀해주신 부분들을 모두 반영하였고, 테스트 결과는 아래와 같습니다.

```
$ go test ./internal/service/hadoop/hadoop_products_data_source_test.go -v
=== RUN   TestAccDataSourceNcloudHadoopProducts_basic
--- PASS: TestAccDataSourceNcloudHadoopProducts_basic (2.75s)
PASS
ok      command-line-arguments  3.610s
```

jungchanSon committed:
refactor(hadoop): Rename var `productsResp` to `hadoopProductsResp`

youngmn reviewed (approved):
LGTM 👍
